### PR TITLE
[Feature] Export chainable state interfaces from chainable controllers

### LINKED ISSUE #995: [Chained Controllers] Chaining of State Interfaces
The controllers chaining described in [Cascade Control Proposal](https://github.com/ros-controls/roadmap/blob/master/design_drafts/cascade_control.md) is not support only chaining of command interface between controllers.
The functionality is based on the same principles for communication between controllers and hardware, where hardware offers `Command Interfaces` and controllers as loaning them based on the name matching using `Loaned Command Interfaces`.
Since such controllers need additional fields and methods, a new type of base controller class is introduced called `ChainableControllerInterface`.
This interface can be used for both classical controllers and chainable controllers. So if you are uncertain if your controller should be chainable in the future, you can implement this base class and simply export an empty list of reference interfaces and disable switching to chained mode.
Those interfaces are called `Reference Interfaces` in the controllers (still have the same type of `CommandInterfaces`).

`Reference Interfaces` are exported when a `Chainable Controller` is configured and stored in the resource manager. When a controller is to be activated, controller manager check if requested command interfaces are from another chainable controllers and if so, it confirms that the following controller is active.

**NOTE** currently the controllers have to be loaded in order from the beginning toward the end of the chain 

For chaining state interfaces, the same logic can be used:

1. Extend existing `ChainableControllerInterface` with method for exporting `Estimated Interfaces` (type `StateInterface`)
2. Get those exported interfaces when configuring controllers in controller manager ([somewhere here](https://github.com/ros-controls/ros2_control/blob/master/controller_manager/src/controller_manager.cpp#L505-L521)
3. Understand the logic of current controller switching for the chained controllers and extend it to support also chaining of state interfaces – begins here](https://github.com/ros-controls/ros2_control/blob/master/controller_manager/src/controller_manager.cpp#L539) --> basically the only trick is to properly fill the activation and deactivation lists.
4. Dr. Denis recommends to actually write first the tests with the controller chain setup (a figure is also helpful) – because then it will be clear which situations should be covered. [Check the tests here](https://github.com/ros-controls/ros2_control/blob/master/controller_manager/test/test_controllers_chaining_with_controller_manager.cpp).

### PR DESCRIPTION
**Pull request by saikishor**
Hello,

This PR adds a new feature to the chainable controllers so that they are able to export the expected interfaces (in the form of state interfaces) that can be used by multiple controllers at the same time. Please take a look at the following image for better understanding:

ros2_chaining_controllers(2) drawio

There could be a use case, where you have a controller that wants to share information with multiple controllers, for instance, a controller like a state estimator, then using this feature you will be able to share the state as estimated interfaces so that other following controllers can use it as a state interface.

Closes https://github.com/ros-controls/ros2_control/issues/995

### CODE REVIEW

**Comment by destogl** [comment]
> There could be a use case, where you have a controller that wants to share information with multiple controllers, for instance, a controller like a state estimator, then using this feature you will be able to share the state as estimated interfaces so that other following controllers can use it as a state interface.

Can you add in the figure connection between robot localization and position tracking? Also, I would remove `odom_publisher` no need to have it, since controllers can publish things directly.

**Comment by saikishor** [comment]
> Can you add in the figure connection between robot localization and position tracking? Also, I would remove `odom_publisher` no need to have it, since controllers can publish things directly.

Thank you @destogl for the reply. Sure, I can add the connection from the `robot_localization` to `position_tracking`. I gree that the `odom_publisher` is not needed as it is very evident, however, I would leave it for the clarity of the same interface resource being utilized by multiple controllers. What do you think? Do you still want me to remove it?.

File: controller_manager/src/controller_manager.cpp
Comment (olivier-stasse) [inline controller_manager/src/controller_manager.cpp:1217]: estimated -> state

File: controller_manager/src/controller_manager.cpp
Comment (olivier-stasse) [inline controller_manager/src/controller_manager.cpp:1218]: estimated->state

**Review by olivier-stasse — APPROVED** [review]
Thanks a lot for this contribution.
The tests are very much appreciated.
I tested the branch and the tests on rolling and this is working.

My main suggestion will be to not use estimated_interfaces which I found misleading.
MPC could provide gains, future state, so for me we should either having a class of interfaces and then derived from it, or stick to state_interface terminology.

**Comment by destogl** [comment]
> What do you think? Do you still want me to remove it?.

We should work on a realistic example, i.e. have real use-case in docs. Otherwise people start to re-create whatever they see.

**Comment by olivier-stasse** [comment]
> > What do you think? Do you still want me to remove it?.
> 
> We should work on a realistic example, i.e. have real use-case in docs. Otherwise people start to re-create whatever they see.

Dear @destogl  to give more concrete details, in humanoid robotics a recent nice explanation of the most common control design pattern is described here:
https://hal.archives-ouvertes.fr/hal-01875387/document
In Fig. 2 you will find a DCM and a ZMP observer which are used by two controllers.
The video of the real experiment is available here:
https://scaron.info/publications/icra-2019.html

You will see that this was implemented in HRP-4 using OpenRTM and mc_rtc (two others conrol framework using various middleware).
We reimplemented 4 or 5 variations in my lab with our own custom architecture.

Another real life example tested by one of our master student similar to the one given by @saikishor can be found in this document https://hal.laas.fr/hal-02964034v1. In p15 you will find a more detailed and complex example of what @saikishor is describing. The state estimator is used for:
* the DCM Controller
* the CoM admittance controller
* the Generalized Inverted Kinematics

We reimplemented the same design pattern for a conference paper in Humanoids 2022 https://hal.laas.fr/LAAS-GEPETTO/hal-03452196 mixing our new in house controller and PAL robotics Inverse dynamics (see Fig.3 of this paper) . It allowed us to perform torque control on TALOS (https://youtu.be/NdZpPIR-hKM).

During ICRA 2023 a discussion with another group (DFKI) reimplemented another variation of the same design pattern. 
We still have issue with some of the state estimators. This has a drastic impact on the quality of the control.
Being able to change it and handling the impact on the other controllers.

I hope this convince you that this is a realistic example, or at least one of the humanoid robotics community.

My hope is that by using ros2_control chainable architecture  we will be able to exchange some of these control blocks.

IMHO the examples given in the previously cited paper are more complex than the theoretical one provided by @saikishor.
But  I can create a PR providing links to the previously cited examples to give more weight to the argument.

**Comment by saikishor** [comment]
Hello @olivier-stasse,

Thank you so much for your review.

**Comment by saikishor** [comment]
> We should work on a realistic example, i.e. have real use-case in docs. Otherwise people start to re-create whatever they see.

Hello @destogl,

I'm sorry that I couldn't reply to you sooner as I was busy with ICRA conference. I agree with you that others might recreate things if the examples are not so realistic enough. However, I would like to have one more in the diagram, to clearly show the paradigm of the newly added estimated interfaces as we wanted them to have a realistic example. How about changing the name from `odom_publisher` to `slip_detection`? or any other name that's meaningful and makes sense? Please let me know.

Regarding the review of @olivier-stasse, what do we do with the naming `estimated` to `state`?, can we have a discussion on it?. I will be waiting for your review.

Best Regards,
Sai

**Comment by christophfroehlich** [comment]
> Regarding the review of @olivier-stasse, what do we do with the naming `estimated` to `state`?, can we have a discussion on it?. I will be waiting for your review.

I agree with Olivier that we shouldn't use `estimated` here. From my understanding of control theory this is clearly related to the estimation of a dynamical system's state. In the chained control loop there could be some kinematic calculations in the backward path (odometry,..), or a block performing signal processing (filtering, outlier detection..) -> no one would call them estimator. 

After merging #985 we should also add docs about the state interfaces in there.

**Comment by jakewelde** [comment]
I just wanted to echo support for this feature improvement. I'm using ros2_control for aerial vehicles, and have a situation in which one hardware component provides state_interfaces for the measurements from an IMU (gyroscope and accelerometer) and another provides lower bandwidth pose estimates (e.g. a motion capture system, a 3rd party visual-inertial odometry system). I would like to run an "estimator" controller that is then chained with the actual controller that takes the system's kinematic states as its state input, but at present the flow of information in chained controllers is essentially one-way.

If this gets merged in, I will definitely make use of it! Thanks!

**Comment by jakewelde** [comment]
Hey @saikishor, thanks again for your hard work on these features! I was just wondering if you have a sense of the approximate horizon over which exporting state interfaces from controllers is likely to get merged in? I'd love to make use of this functionality, but may make some different architectural decisions if this is slated more towards the long term development plan. Any general guidance you can offer is a great help. Thank you!

**Comment by JohannesPankert** [comment]
Same here, this is a feature that we very much need and we are happy if this will be merged soon.

**Comment by saikishor** [comment]
Hello @jakewelde and @JohannesPankert!

I appreciate your interest in this feature. I hope this gets merged soon too. This is under review currently. Most likely it will be released before the next ROS2 LTS release. @destogl is presently reviewing the PR. In the meantime, once the PR about sorting the controller while configuring is merged: https://github.com/ros-controls/ros2_control/pull/1107, I will proceed to fix the controller sorting to also consider the controllers that expose the state interfaces. Right now, this part of it is broken. Once this is done, this PR is good to go.

Thank you for understanding,

Best Regards,
Sai

**Comment by jakewelde** [comment]
Thanks very much for the updates @saikishor, and for your contributions! Looking forward to using this feature.

**Comment by jukindle** [comment]
Same here! Currently to achieve this separation of a state estimator controller in ros2control, we have to create a fake actuator holding the state estimate in the hardware interface which is then written to by the state estimator controller and read from by all other controllers - this results in the state estimate potentially being delayed by one update call.. I really hope this merge request makes it into Jazzy

**Review by christophfroehlich — APPROVED** [review]
The changes look fine for me, and reflects the results of the discussions afaik.

Could you please update the [docs](https://github.com/ros-controls/ros2_control/blob/master/controller_manager/doc/controller_chaining.rst) as well? Or should we open a follow-up issue?

File: controller_interface/include/controller_interface/controller_interface.hpp
Comment (bmagyar) [inline controller_interface/include/controller_interface/controller_interface.hpp:46]:
```suggestion
   * A non-chainable controller has no state interfaces.
```
Wondering is this wording is better?

File: controller_interface/test/test_chainable_controller_interface.cpp
Comment (bmagyar) [inline controller_interface/test/test_chainable_controller_interface.cpp:47]:
```suggestion
  ASSERT_THAT(internal_state_interfaces, SizeIs(1));
```
I know I'm nitpicking but let's use the advanced features of gmock. The difference is that you get much more informative prints when something isn't right.

You may need to do `using testing::SizeIs;` at the top if it's not there yet

File: controller_interface/test/test_chainable_controller_interface.cpp
Comment (bmagyar) [inline controller_interface/test/test_chainable_controller_interface.cpp:92]:
```suggestion
  ASSERT_THAT(internal_state_interfaces, IsEmpty());
```
I know I'm nitpicking but let's use the advanced features of gmock. The difference is that you get much more informative prints when something isn't right.

You may need to do `using testing::IsEmpty()` at the top if it isn't there yet

File: hardware_interface/src/resource_manager.cpp
Comment (bmagyar) [inline hardware_interface/src/resource_manager.cpp:467]: I think it may be better to not duplicate docstrings in the cpp. I know it's handy when developing as a reminder and also when reading but I think the header file is the place where this should be and comments in the CPP file should be reserved for implementation-specific notes if applicable

Comment (bmagyar) [reply hardware_interface/src/resource_manager.cpp:467]:
```suggestion
```

**Review by bmagyar — CHANGES_REQUESTED** [review]
Just small comments

File: hardware_interface/src/resource_manager.cpp
Comment (saikishor) [reply hardware_interface/src/resource_manager.cpp:467]: @bmagyar I just verified the code, I had to add it in the cpp file because these methods belong to the ResourceStorage class and this is defined inside the cpp file and there is no declaration of the methods in the header files.

What should I do now in that case?

File: controller_interface/include/controller_interface/controller_interface.hpp
Comment (saikishor) [reply controller_interface/include/controller_interface/controller_interface.hpp:46]:
```suggestion
   * A non-chainable controller doesn't export any state interfaces.
```

**Review by bmagyar — APPROVED** [review]
Thanks for the great work!

**Comment by jakewelde** [comment]
Hey folks! Thanks again for all your hard work on this. I was just curious, do you know if this feature improvement is still expected to be released in conjunction with ROS 2 Jazzy? We are still quite eager to make use of it! Thanks.

**Comment by christophfroehlich** [comment]
> Hey folks! Thanks again for all your hard work on this. I was just curious, do you know if this feature improvement is still expected to be released in conjunction with ROS 2 Jazzy? We are still quite eager to make use of it! Thanks.

yes, it is still the plan to do so. You may help by compiling this branch from source, test it, and write a review.

File: controller_interface/include/controller_interface/controller_interface_base.hpp
Comment (destogl) [inline controller_interface/include/controller_interface/controller_interface_base.hpp:267]: Note to myself: "Why do we need this?"

File: controller_interface/src/chainable_controller_interface.cpp
Comment (destogl) [inline controller_interface/src/chainable_controller_interface.cpp:55]:
```suggestion
```

File: controller_manager/src/controller_manager.cpp
Comment (destogl) [inline controller_manager/src/controller_manager.cpp:128]:
```suggestion
      "Required interface '%s' with prefix '%s' is not a chain interface.",
```

File: controller_manager/src/controller_manager.cpp
Comment (destogl) [inline controller_manager/src/controller_manager.cpp:1704]: Why `estimate`, wouldn't be `interface` more appropriate?

**Review by destogl — CHANGES_REQUESTED** [review]
I have a proposal how to simplify this PR in general. I don't see an need to have additionally `use_references_from_subsribers` type of code as this is automatically acivated when a controller is in chined mode. On the other hand, when controller provides state interface for another controller, than we don't actually have to track if controller is in chainable mode or not, as for the providing controller nothing changes if someone is listening or not on his outputs. This means that we write to the state interfaces all the time and the following controller (that uses that data) should take care that the controller is activated.

This would simplifiy logic in the CM and also reduce interfaces of controllers. And of course provide some code clarity, because now it is getting confusing with "chained" / "referenced from subscribers".

This mena

File: controller_manager/include/controller_manager/controller_manager.hpp
Comment (saikishor) [reply controller_manager/include/controller_manager/controller_manager.hpp:400]: Yes ofcourse. The idea is the same to have it available always, but the problem here is when the state interfaces are being used the controller should report it as it being in the chained mode or not. I thought it would make sense for it to be reported chained mode but then we have a separate method to manage the switching of references between subscribers or interfaces

File: controller_manager/src/controller_manager.cpp
Comment (saikishor) [reply controller_manager/src/controller_manager.cpp:136]: Basically, it checks if the interface matches with the given controller name. I can merge it with the next method as a lamba, as this is the only one being used

File: controller_manager/src/controller_manager.cpp
Comment (saikishor) [reply controller_manager/src/controller_manager.cpp]: This is needed to do some internal checking, maybe this can be replaced with new ControllerChainSpec

**Review by saikishor — COMMENTED** [review]
@destogl Thanks for the review. I've addressed most of them and responded to some comments

**Comment by saikishor** [comment]
@destogl I've addressed the changes as per your suggestion. Now the usage of the exported state interfaces doesn't put the controller in chained mode. It will be only be in chained mode when it's reference interfaces are used by other controller to command. It would be great. if you can take a look :)

**Review by christophfroehlich — CHANGES_REQUESTED** [review]
Requesting docs ([here](https://control.ros.org/master/doc/ros2_control/controller_manager/doc/controller_chaining.html) and into the release notes) :innocent: or at least a follow-up issue

**Comment by saikishor** [comment]
> Requesting docs ([here](https://control.ros.org/master/doc/ros2_control/controller_manager/doc/controller_chaining.html) and into the release notes) 😇 or at least a follow-up issue

@christophfroehlich The docs and the release notes are updated :)

Thank you!

File: controller_manager/doc/controller_chaining.rst
Comment (christophfroehlich) [inline controller_manager/doc/controller_chaining.rst:62]:
```suggestion
For simplicity reasons, it is assumed for now that controller's all reference interfaces are used by other controller. However, the state interfaces exported by the controller, can be used by multiple controllers at the same time and with the combination they want.
```

File: controller_manager/doc/controller_chaining.rst
Comment (christophfroehlich) [inline controller_manager/doc/controller_chaining.rst:61]:
```suggestion
This method should be implemented for each controller that **can be preceded** by another controller exporting all the reference/state interfaces.
```

File: controller_manager/doc/controller_chaining.rst
Comment (christophfroehlich) [inline controller_manager/doc/controller_chaining.rst:65]:
```suggestion
The ``ChainableController`` base class implements ``void set_chained_mode(bool activate)`` that sets an internal flag that a controller is used by another controller (in chained mode) and calls ``virtual void on_set_chained_mode(bool activate) = 0`` that implements controller's specific actions when chained mode is activated or deactivated, e.g., deactivating subscribers.
```

**Review by christophfroehlich — CHANGES_REQUESTED** [review]
Am I missing something, or should we briefly introduce what a state interface or reference interface is?
We introduced "reference" [here](https://github.com/ros-controls/roadmap/blob/master/design_drafts/design_of_controllers_and_their_status.md) but in the context of chainable controllers, this distinction is not yet very clear imho.

**Comment by saikishor** [comment]
> Am I missing something, or should we briefly introduce what a state interface or reference interface is? We introduced "reference" [here](https://github.com/ros-controls/roadmap/blob/master/design_drafts/design_of_controllers_and_their_status.md) but in the context of chainable controllers, this distinction is not yet very clear imho.

@christophfroehlich I've added a bit more of the documentation! Thank you!

**Review by christophfroehlich — DISMISSED** [review]

File: controller_interface/src/chainable_controller_interface.cpp
Comment (bmagyar) [inline controller_interface/src/chainable_controller_interface.cpp:80]: The code to generate the *almost* one and only implementation that meets this criteria in a meaningful way takes a vector of strings, pairs them with the values in `state_interfaces_values_` and creates a vector of `StateInterface`s. I think we should not make people copy that piece of code around but instead generate it right here, with the right prefix and everything.

Comment (bmagyar) [reply controller_interface/src/chainable_controller_interface.cpp:80]: In my mind both values and names should be set up in a single place, either the constructor or `init` or `configure`, the state interface export can be implemented by us.

**Review by bmagyar — CHANGES_REQUESTED** [review]

**Review by destogl — CHANGES_REQUESTED** [review]
Just two more comments.

File: controller_interface/src/chainable_controller_interface.cpp
Comment (saikishor) [reply controller_interface/src/chainable_controller_interface.cpp:80]: I agree with you about having methods that do it automatically for the user. I would make it optional as in the user can decide to use those helper methods or not. I personally don't like to force users to use this specific data type. For instance, one of my state interfaces is the position of the base of the robot, and sometimes it is better to have it `Eigen::Vector3d` or `geometry_msgs/Point`. This would give some flexibility to the user in my opinion.

If it is only the above position it is not a problem, but if he has orientation and all joints state etc, then it's a huge chunk of data and it gets messy when the user wants to remap the data into his own struct.

What do you think about it?

File: controller_interface/include/controller_interface/chainable_controller_interface.hpp
Comment (saikishor) [reply controller_interface/include/controller_interface/chainable_controller_interface.hpp:80]: Are you talking about this `on_export_state_interfaces` method? or the `export_state_interfaces` method?

If it is `export_state_interfaces` method, then as you mentioned it already exists in the `controller_interface_base.hpp` and a default implementation is done inside `controller_interface.cpp` as it is done similar to the  `export_reference_interfaces`

**Review by MaximilienNaveau — APPROVED** [review]

File: controller_interface/src/chainable_controller_interface.cpp
Comment (bmagyar) [reply controller_interface/src/chainable_controller_interface.cpp:80]: What if we left this check but had the suggested default implementation too? Still erring on the side of not making people copy&paste the same boilerplate code all the time.

Comment (saikishor) [reply controller_interface/src/chainable_controller_interface.cpp:80]: Sure, we can do that. I can do the boilerplate part, and we can still do error on this side. I like it

Comment (saikishor) [reply controller_interface/src/chainable_controller_interface.cpp:80]: I've just implemented it by adding a member variable and the user fills in the list of variable names to be exported in the configure stage etc and it will done automatically. 
Take a look at the implementation and let me know.

Thank you :)

Comment (bmagyar) [reply controller_interface/src/chainable_controller_interface.cpp:80]: Thank you!

**Review by christophfroehlich — APPROVED** [review]

**Review by bmagyar — APPROVED** [review]

**Review by destogl — APPROVED** [review]
Thanks for following up on this!!!

**Comment by saikishor** [comment]
> Any chance this is backported to humble?

@InigoMoreno currently, we don't have plans to backport this feature to Humble as it breaks both API and ABI

**Comment by christophfroehlich** [comment]
@InigoMoreno are you aware that the rolling branch is compatible with humble if you compile from source?

**Comment by InigoMoreno** [comment]
> @InigoMoreno are you aware that the rolling branch is compatible with humble if you compile from source?

I was not aware of that, thanks for telling me!

**Comment by christophfroehlich** [comment]
see control.ros.org, for which repos file to use.

**Pull request merged**
